src/expr/src/expr/expr_udf.rs

```diff
@@ -52,15 +52,21 @@ impl Expression for UdfExpression {
             columns.push(array.as_ref().into());
         }
         let opts =
-            arrow_array::RecordBatchOptions::default().with_row_count(Some(input.cardinality()));
+            arrow_array::RecordBatchOptions::default().with_row_count(Some(input.capacity()));
         let input =
             arrow_array::RecordBatch::try_new_with_options(self.arg_schema.clone(), columns, &opts)
                 .expect("failed to build record batch");
         let output = self.client.call(&self.identifier, input).await?;
-        let arrow_array = output
-            .columns()
-            .get(0)
-            .ok_or(risingwave_udf::Error::NoColumn)?;
+        if output.num_rows() != vis.len() {
+            bail!(
+                "UDF returned {} rows, but expected {}",
+                output.num_rows(),
+                vis.len(),
+            );
+        }
+        let Some(arrow_array) = output.columns().get(0) else {
+            bail!("UDF returned no columns");
```

Comment:
When an expression returns an Error, we will put an array of `NULL`s there, right? 

If so, it may sounds a little bit violent. For example, is it possible to fallback to a row-by-row mode? so that at least we can fill these normal values.

Reply:
I guess so. In this case it will fallback in an upper-level function:
https://github.com/risingwavelabs/risingwave/blob/db072cb9baebe397036bf19b5afbc238b819812b/src/expr/src/expr/mod.rs#L122

Although it's not efficient to evaluate row-by-row in UDF.